pkg/packetfilter/nftables/nftables.go

```diff
@@ -0,0 +1,535 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nftables
+
+import (
+	"context"
+	"slices"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/log"
+	"github.com/submariner-io/submariner/pkg/packetfilter"
+	"k8s.io/utils/ptr"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/knftables"
+)
+
+const (
+	/* Single table named 'submariner' is used for nftables configuration.*/
+	submarinerTable = "submariner"
+)
+
+var (
+	iphookChainTypeToNftablesType = map[packetfilter.ChainType]knftables.BaseChainType{
+		packetfilter.ChainTypeFilter: knftables.FilterType,
+		packetfilter.ChainTypeRoute:  knftables.RouteType,
+		packetfilter.ChainTypeNAT:    knftables.NATType,
+	}
+
+	iphookChainHookToNftablesHook = map[packetfilter.ChainHook]knftables.BaseChainHook{
+		packetfilter.ChainHookPrerouting:  knftables.PreroutingHook,
+		packetfilter.ChainHookInput:       knftables.InputHook,
+		packetfilter.ChainHookForward:     knftables.ForwardHook,
+		packetfilter.ChainHookOutput:      knftables.OutputHook,
+		packetfilter.ChainHookPostrouting: knftables.PostroutingHook,
+	}
+
+	iphookChainTypeToNftablesBasicPriority = map[packetfilter.ChainType]knftables.BaseChainPriority{
+		packetfilter.ChainTypeFilter: knftables.FilterPriority,
+		packetfilter.ChainTypeRoute:  knftables.ManglePriority,
+		packetfilter.ChainTypeNAT:    knftables.DNATPriority,
+	}
+
+	ruleActionToStr = map[packetfilter.RuleAction][]string{
+		packetfilter.RuleActionAccept: {"accept"},
+		packetfilter.RuleActionMss:    {"tcp", "option", "maxseg"},
+		packetfilter.RuleActionMark:   {"meta", "mark"},
+		packetfilter.RuleActionSNAT:   {"snat"},
+		packetfilter.RuleActionDNAT:   {"dnat"},
+		packetfilter.RuleActionJump:   {"jump"},
+	}
+
+	logger = log.Logger{Logger: logf.Log.WithName("NFTables")}
+)
+
+type RuleSpec []string
+
+func (r RuleSpec) String() string {
+	return strings.Join(r, " ")
+}
+
+type packetFilter struct {
+	nftables knftables.Interface
+}
+
+func New() (packetfilter.Driver, error) {
+	nft, err := knftables.New(knftables.IPv4Family, submarinerTable)
+	if err != nil {
+		return nil, errors.Wrap(err, "error creating knftables")
+	}
+
+	return &packetFilter{
+		nftables: nft,
+	}, nil
+}
+
+func (p *packetFilter) ChainExists(_ packetfilter.TableType, chain string) (bool, error) {
+	return p.chainExists(chain)
+}
+
+func (p *packetFilter) chainExists(chain string) (bool, error) {
+	chainsList, err := p.nftables.List(context.TODO(), "chains")
+	if err != nil && !knftables.IsNotFound(err) {
+		return false, errors.Wrap(err, "error listing chains")
+	}
+
+	return slices.Contains(chainsList, chain), nil
+}
+
+func (p *packetFilter) AppendUnique(_ packetfilter.TableType, chain string, rule *packetfilter.Rule) error {
+	return p.insertRuleAtPosition(chain, rule, -1)
```

Comment:
The knftables API doesn't ensure uniqueness so perhaps we should implement that here like we did for `PrependUnique`.

Reply:
In case of AppendUnique , p.insertRuleAtPosition function is called with position = -1, which means we'll use tx.Add(&knftRule) without specifying index .
AFAIU, from knftables doc [1] that should ensure uniqueness.

[1]
// Add adds an "nft add" operation to tx, ensuring that obj exists by creating it if it
// did not already exist. (If obj is a Rule, it will be appended to the end of its chain,
// or else added after the Rule indicated by this rule's Index or Handle.)

Reply:
So that's the docs for the general `tx.Add` operation and create-if-not-exists may not apply to all objects. In the fake implementation, it does apply to chains, sets, maps et al. But it doesn't apply to rules (i tried `AppendUnique` plus I can see it in the code).  I think this is b/c rules are a bit different in that they don't have an persistent unique identifier, like a chain or set that has a name. They have a handle but that's not assigned until it's created.  So `knftables` doesn't have a way to check for an existing rule. In fact, I think this [doc](https://github.com/kubernetes-sigs/knftables?tab=readme-ov-file#missing-apis) confirms it: "_Most IPTables libraries have an API for "add this rule only if it doesn't already exist", but that does not seem as useful in nftables (or at least "in nftables as used by Kubernetes-ish components that aren't just blindly copying over old iptables APIs"), because chains tend to have static rules and dynamic sets/maps, rather than having dynamic rules._".  So we're going to have to ensure uniqueness on append. But we can do this in a followup. I have cycles if you want after I add the unit tests.

Reply:
Ack, yep we can have it in a followup PR.
10x